jerry-core/ecma/operations/ecma-promise-object.c

```diff
@@ -609,32 +609,79 @@ ecma_promise_remaining_inc_or_dec (ecma_value_t remaining, /**< the remaining co
 } /* ecma_promise_remaining_inc_or_dec */
 
 /**
- * Native handler for Promise.all Resolve Element Function.
+ * Native handler for Promise.all and Promise.allSettled Resolve Element Function.
  *
  * See also:
  *         ES2015 25.4.4.1.2
  *
  * @return ecma value of undefined.
  */
 ecma_value_t
-ecma_promise_all_handler_cb (ecma_object_t *function_obj_p, /**< function object */
+ecma_promise_all_or_all_settled_handler_cb (ecma_object_t *function_obj_p, /**< function object */
                              const ecma_value_t args_p[], /**< argument list */
                              const uint32_t args_count) /**< argument number */
 {
   JERRY_UNUSED (args_count);
   ecma_promise_all_executor_t *executor_p = (ecma_promise_all_executor_t *) function_obj_p;
+  uint16_t promise_type = executor_p->header.u.class_prop.extra_info;
 
   /* 1 - 2. */
   if (executor_p->index == 0)
   {
     return ECMA_VALUE_UNDEFINED;
   }
 
-  /* 8. */
-  ecma_op_object_put_by_index (ecma_get_object_from_value (executor_p->values),
-                               (uint32_t) (executor_p->index - 1),
-                               args_p[0],
-                               false);
+  if (promise_type == ECMA_PROMISE_ALL_RESOLVE)
+  {
+    /* 8. */
+    ecma_op_object_put_by_index (ecma_get_object_from_value (executor_p->values),
+                                 (uint32_t) (executor_p->index - 1),
+                                 args_p[0],
+                                 false);
+  }
+  else
+  {
+    ecma_object_t *object_prototype_p = ecma_builtin_get (ECMA_BUILTIN_ID_OBJECT_PROTOTYPE);
+    ecma_object_t *obj_p = ecma_create_object (object_prototype_p, 0, ECMA_OBJECT_TYPE_GENERAL);
+    ecma_property_value_t *prop_value_p;
+    prop_value_p = ecma_create_named_data_property (obj_p,
+                                                    ecma_get_magic_string (LIT_MAGIC_STRING_STATUS),
+                                                    ECMA_PROPERTY_CONFIGURABLE_ENUMERABLE_WRITABLE,
+                                                    NULL);
+
+    if (promise_type == ECMA_PROMISE_ALLSETTLED_RESOLVE)
+    {
+      prop_value_p->value = ecma_make_magic_string_value (LIT_MAGIC_STRING_FULFILLED);
+    }
+    else
+    {
+      prop_value_p->value = ecma_make_magic_string_value (LIT_MAGIC_STRING_REJECTED);
+    }
+
+    if (promise_type == ECMA_PROMISE_ALLSETTLED_RESOLVE)
```

Comment:
I would create a local string (name_p), where the `ecma_get_magic_string (LIT_MAGIC_STRING_VALUE)` or `ecma_get_magic_string (LIT_MAGIC_STRING_REASON)` is stored, and call ecma_create_named_data_property only once.